pkg/upload/js.go

```diff
@@ -0,0 +1,376 @@
+				continue
+			}
+
+			// Remove any loaders
+			questionMark := strings.LastIndex(sourcePath, "?")
+			if questionMark-1 > 0 {
+				sourcePath = sourcePath[:questionMark-1]
+			}
+		}
+		if !path.IsAbs(sourcePath) {
+			sourcePath = path.Join(projectRoot, sourcePath)
+		}
+		logger.Debug(fmt.Sprintf("Attempting to read the source %s.", sourcePath))
+		content, err := os.ReadFile(sourcePath)
+		if err != nil {
+			logger.Warn(fmt.Sprintf("Cannot read referenced source file '%s': %s", sourcePath, err))
+			sourcesContent = append(sourcesContent, nil)
+		} else {
+			contentString := string(content)
+			sourcesContent = append(sourcesContent, &contentString)
+		}
+	}
+
+	section["sourcesContent"] = sourcesContent
+	return true
+}
+
+// If the source map doesn't have sourcesContent then attempt to add it. Returns true if a modifcation was performed.
+func AddSources(sourceMapContents map[string]interface{}, sourceMapPath string, projectRoot string, logger log.Logger) bool {
+	// Sources may be in several sections. See https://bit.ly/sourcemap.
+	if sections, exists := sourceMapContents["sections"]; exists {
+		if sources, ok := sections.([]map[string]interface{}); ok {
+			anyModified := false
+			for _, section := range sources {
+				modified := addSourcesContent(section, sourceMapPath, projectRoot, logger)
+				anyModified = modified || anyModified
+			}
+			return anyModified
+		}
+	} else {
+		return addSourcesContent(sourceMapContents, sourceMapPath, projectRoot, logger)
+	}
+	return false
+}
+
+// Attempt to find the bundle path by changing the extension of the source map if the bundle path is not passed in to the command line
+func resolveBundlePath(bundlePath string, sourceMapPath string, logger log.Logger) (string, error) {
+	if bundlePath != "" {
+		if utils.FileExists(bundlePath) {
+			return bundlePath, nil
+		} else {
+			return "", fmt.Errorf("unable to find specified bundle: %s", bundlePath)
+		}
+	}
+
+	withoutSuffix, found := strings.CutSuffix(sourceMapPath, ".map")
+	if !found {
+		return "", nil
+	}
+
+	if utils.FileExists(withoutSuffix) {
+		logger.Info(fmt.Sprintf("Automatically using the bundle at path %s based on stripping the .map suffix.", withoutSuffix))
+		return withoutSuffix, nil
+	}
+	return "", nil
+}
+
+// Upload a single source map
+func uploadSingleSourceMap(
+	bundleUrl string,
+	baseUrl string,
+	bundlePath string,
+	sourceMapPath string,
+	apiKey string,
+	appVersion string,
+	projectRoot string,
+	endpoint string,
+	timeout int,
+	retries int,
+	overwrite bool,
+	dryRun bool,
+	logger log.Logger,
+) {
+
+	sourceMapContents, err := ReadSourceMap(sourceMapPath, logger)
+	if err != nil {
+		logger.Fatal(err.Error())
+	}
+
+	var sourceMapFile server.FileField
+
+	sourceMapModified := AddSources(sourceMapContents, sourceMapPath, projectRoot, logger)
+	if sourceMapModified {
+		encodedSourceMap, err := json.Marshal(sourceMapContents)
+		if err != nil {
+			logger.Fatal(err.Error())
```

Comment:
I think more context here would be useful:

```suggestion
			logger.Fatal(fmt.Sprintf("Failed generate valid source map JSON with original sources added: %s", err.Error()))
```